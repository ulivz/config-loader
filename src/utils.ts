/**
 * Forked from https://github.com/egoist/bundle-require/blob/main/src/index.ts
 * Thanks for the great work from @egoist.
 */
import fs from 'fs';
import path from 'path';

export function getPackagesFromNodeModules(dir = 'node_modules') {
  const result: string[] = [];
  const names = fs.existsSync(dir) ? fs.readdirSync(dir) : [];
  for (const name of names) {
    if (name[0] === '@') {
      try {
        const subnames = fs.readdirSync(path.join(dir, name));
        for (const subname of subnames) {
          result.push(`${name}/${subname}`);
        }
      } catch (error) {
        result.push(name);
      }
    } else {
      result.push(name);
    }
  }
  return result;
}
