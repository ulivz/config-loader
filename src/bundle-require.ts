/**
 * Forked from https://github.com/egoist/bundle-require/blob/main/src/index.ts
 * Thanks for the great work from @egoist.
 */
import fs from 'fs';
import path from 'path';
import { build, Loader, Plugin, BuildOptions } from 'esbuild';
import { getPackagesFromNodeModules } from './utils';

const JS_EXT_RE = /\.(mjs|cjs|ts|js|tsx|jsx)$/;

function inferLoader(ext: string): Loader {
  if (ext === '.mjs' || ext === '.cjs') return 'js';
  return ext.slice(1) as Loader;
}

export interface Options {
  /**
   * The filepath to bundle and require
   */
  filepath: string;
  /**
   * The `require` function that is used to load the output file
   * Default to the global `require` function
   * This function can be asynchronous, i.e. returns a Promise
   */
  // eslint-disable-next-line no-unused-vars
  require?: (outfile: string) => any;
  /**
   * esbuild options
   */
  esbuildOptions?: BuildOptions;
  /**
   * esbuild plugin
   */
  esbuildPlugins?: Plugin[];
  /**
   * Get the path to the output file
   * By default we simply replace the extension with `.bundled.cjs`
   */
  // eslint-disable-next-line no-unused-vars
  getOutputFile?: (filepath: string) => string;
  /**
   * Whether the keep the temporary file.
   */
  preserveTemporaryFile?: boolean;
}

const defaultGetOutputFile = (filepath: string) =>
  filepath.replace(JS_EXT_RE, '.bundled.cjs');

export async function bundleRequire(options: Options) {
  if (!JS_EXT_RE.test(options.filepath)) {
    throw new Error(`${options.filepath} is not a valid JS file`);
  }

  const getOutputFile = options.getOutputFile || defaultGetOutputFile;
  const outfile = getOutputFile(options.filepath);

  const packageNames = getPackagesFromNodeModules();

  await build({
    entryPoints: [options.filepath],
    outfile,
    format: 'cjs',
    platform: 'node',
    bundle: true,
    ...options.esbuildOptions,
    plugins: [
      ...(options.esbuildPlugins || []),
      {
        name: 'replace-path',
        setup(ctx) {
          ctx.onResolve({ filter: /.*/ }, (args) => {
            const isPackage = packageNames.some(
              (name) => args.path === name || args.path.startsWith(`${name}/`)
            );
            if (isPackage) {
              return {
                path: args.path,
                external: true,
              };
            }
          });

          ctx.onLoad({ filter: JS_EXT_RE }, async (args) => {
            const contents = await fs.promises.readFile(args.path, 'utf-8');
            return {
              contents: contents
                .replace(/\b__filename\b/g, JSON.stringify(args.path))
                .replace(
                  /\b__dirname\b/g,
                  JSON.stringify(path.dirname(args.path))
                )
                .replace(
                  /\bimport\.meta\.url\b/g,
                  JSON.stringify(`file://${args.path}`)
                ),
              loader: inferLoader(path.extname(args.path)),
            };
          });
        },
      },
    ],
  });

  let mod: any;
  const req = options.require || require;
  try {
    mod = await req(outfile);
  } finally {
    if (!options.preserveTemporaryFile) {
      // Remove the outfile after executed
      await fs.promises.unlink(outfile);
    }
  }

  return mod;
}
