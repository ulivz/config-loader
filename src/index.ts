/**
 * Module dependencies
 */
import * as fs from 'fs';
import * as path from 'path';
import { bundleRequire, Options } from './bundle-require';

export interface IConfigLoaderOptions {
  cwd: string;
  stopDir?: string;
  configKey: string;
  configFile?: string | string[];
  configSuffix?: string;
  esbuildOptions?: Options['esbuildOptions'];
  preserveTemporaryFile?: boolean;
}

export interface IConfigLoaderResult<T> {
  path?: string;
  data: T;
}

export async function loadConfig<T = Record<string, unknown>>(
  opts: IConfigLoaderOptions
): Promise<IConfigLoaderResult<T>> {
  const {
    cwd = process.cwd(),
    configKey,
    configFile,
    configSuffix = 'config',
    esbuildOptions = {},
  } = opts;

  const files = Array.isArray(configFile)
    ? configFile
    : typeof configFile === 'string'
    ? [configFile]
    : [
        `${configKey}.${configSuffix}.ts`,
        `${configKey}.${configSuffix}.js`,
        `.${configKey}rc.ts`,
        `.${configKey}rc.js`,
      ];

  for (const file of files) {
    const filepath = path.resolve(cwd, file);
    if (fs.existsSync(filepath)) {
      const m = await bundleRequire({
        filepath,
        preserveTemporaryFile: opts.preserveTemporaryFile,
        esbuildOptions: {
          bundle: false,
          ...esbuildOptions,
          target: 'node12',
        },
      });
      const data = m.default || m;
      return {
        path: filepath,
        data,
      };
    }
  }

  return {
    path: undefined,
    data: {} as T,
  };
}
