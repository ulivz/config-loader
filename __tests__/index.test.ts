
import fs from 'fs';
import path from 'path';
import { loadConfig, IConfigLoaderResult } from '../src';

const useScene = (scene: string) => path.join(__dirname, 'fixtures', scene)

const expectResultToMatchSnapshot = (result: IConfigLoaderResult<unknown>) => {
  if (result.path) {
    expect(path.isAbsolute(result.path)).toBe(true);
    const relativePath = path.relative(__dirname, result.path);
    result.path = path.join('/<AbsTestRoot>', relativePath);
  }
  expect(result).toMatchSnapshot();
}

describe('loadConfig', () => {
  it('js', async () => {
    const config = await loadConfig({
      cwd: useScene('js'),
      configKey: 'foo',
    })
    expectResultToMatchSnapshot(config)
  })

  it('ts', async () => {
    const config = await loadConfig({
      cwd: useScene('ts'),
      configKey: 'foo',
    })
    expectResultToMatchSnapshot(config)
  })

  it('custom config', async () => {
    const config = await loadConfig({
      cwd: useScene('custom'),
      configKey: 'foo',
      configFile: 'foo.config.custom.ts',
    })
    expectResultToMatchSnapshot(config)
  })

  it('custom config with array', async () => {
    const config = await loadConfig({
      cwd: useScene('custom'),
      configKey: 'foo',
      configFile: [
        'bar.config.js',
        'bar.config.ts',
        'foo.config.custom.js',
        'foo.config.custom.ts'
      ],
    })
    expectResultToMatchSnapshot(config)
  })

  it('config suffix', async () => {
    const config = await loadConfig({
      cwd: useScene('config-suffix'),
      configKey: 'foo',
      configSuffix: 'generator',
    })
    expectResultToMatchSnapshot(config)
  })

  it('not-existed', async () => {
    const config = await loadConfig({
      cwd: useScene('not-existed'),
      configKey: 'foo',
      configSuffix: 'generator',
    })
    expectResultToMatchSnapshot(config)
  })

  it('absolute path', async () => {
    const configFileWithAbsPath =  path.join(useScene('speedy'), 'speedy.config.ts')
    const config = await loadConfig({
      cwd: useScene('speedy'),
      configKey: 'speedy',
      configFile: configFileWithAbsPath,
    })
    expectResultToMatchSnapshot(config)
  })

  it('preserveTemporaryFile', async () => {
    const tempFile = path.join(useScene('speedy'), 'speedy.config.bundled.cjs')
    const config = await loadConfig({
      cwd: useScene('speedy'),
      configKey: 'speedy',
      preserveTemporaryFile: true,
    })
    expect(fs.existsSync(tempFile)).toBe(true);
  })

  it('tsconfig target environment was set to `es5`', async () => {
    const config = await loadConfig({
      cwd: useScene('target-es5'),
      configKey: 'foo',
    })
    expectResultToMatchSnapshot(config)
  })  
  
  it('import file', async () => {
    const config = await loadConfig({
      cwd: useScene('import-file'),
      configKey: 'foo',
      preserveTemporaryFile: true
    })
    expectResultToMatchSnapshot(config)
  })
})
