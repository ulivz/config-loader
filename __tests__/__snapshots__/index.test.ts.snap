// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfig absolute path 1`] = `
Object {
  "data": Object {},
  "path": undefined,
}
`;

exports[`loadConfig config suffix 1`] = `
Object {
  "data": Object {
    "a": 1,
  },
  "path": "/<AbsTestRoot>/fixtures/config-suffix/foo.generator.ts",
}
`;

exports[`loadConfig custom config 1`] = `
Object {
  "data": Object {
    "a": 1,
  },
  "path": "/<AbsTestRoot>/fixtures/custom/foo.config.custom.ts",
}
`;

exports[`loadConfig custom config with array 1`] = `
Object {
  "data": Object {
    "a": 1,
  },
  "path": "/<AbsTestRoot>/fixtures/custom/foo.config.custom.ts",
}
`;

exports[`loadConfig import file 1`] = `
Object {
  "data": Object {
    "input": Object {
      "a": "b",
    },
  },
  "path": "/<AbsTestRoot>/fixtures/import-file/foo.config.ts",
}
`;

exports[`loadConfig js 1`] = `
Object {
  "data": Object {
    "a": 1,
  },
  "path": "/<AbsTestRoot>/fixtures/js/foo.config.js",
}
`;

exports[`loadConfig not-existed 1`] = `
Object {
  "data": Object {},
  "path": undefined,
}
`;

exports[`loadConfig ts 1`] = `
Object {
  "data": Object {
    "a": 1,
  },
  "path": "/<AbsTestRoot>/fixtures/ts/foo.config.ts",
}
`;

exports[`loadConfig tsconfig target environment was set to \`es5\` 1`] = `
Object {
  "data": Object {
    "a": 1,
    "b": [Function],
  },
  "path": "/<AbsTestRoot>/fixtures/target-es5/foo.config.ts",
}
`;
